---
title: "sctld_analysis_GO_mkd"
author: "Melissa Drown"
date: "11/19/2020"
output: html_document
---

```{r}
library(tximport)
library(DESeq2)
library(tidyverse)
library(tximport)
library(ComplexHeatmap)
#library(DEGreport)
library(limma)
library(edgeR)
library(pheatmap)
library(topGO)
library(ggplot2)
library(gridExtra)
library(reshape2)
#library(mixOmics)
library(RColorBrewer)
library(VennDiagram)
library(devtools)
library(factoextra)
library(topGO)
```

# Gene annotation of ofav
```{r}
setwd("~/Desktop")

ofav_DEG_GLM_treatment <- read.csv("ofav_DEG_GLM_treatment.csv")

ofav_emapper_prot <-read.csv("ofav.protein.emapper.csv")

ofav_prot <- read.csv("ofav_proteins_13173_311351.csv")

```

```{r}
names(ofav_DEG_GLM_treatment)[names(ofav_DEG_GLM_treatment) == "X"] <- "locus"
```


```{r}
ofav_annot_DEG <- left_join(ofav_DEG_GLM_treatment, ofav_prot, by="locus")
```

```{r}
ofav_annot_DEG_emap <- left_join(ofav_annot_DEG, ofav_emapper_prot, by="protein_product")
```

```{r}
#write.csv(ofav_annot_DEG_emap, "ofav_annot_DEG_emap.csv")
```

# volcano plots of DEG from GLM w/treatment
Ref: https://www.rpubs.com/Knight95/volcano
```{r}
# set up data - FDR
ofav_DEG_GLM_treatment$expression = ifelse(ofav_DEG_GLM_treatment$PAdju_BH < 0.05 & abs(ofav_DEG_GLM_treatment$logFC) >= 1, 
                     ifelse(ofav_DEG_GLM_treatment$logFC> 1 ,'Up','Down'),
                     'Stable')

#mcav_DEG_GLM_treatment$expression = ifelse(mcav_DEG_GLM_treatment$PAdju_BH < 0.05 & abs(mcav_DEG_GLM_treatment$logFC) >= 1, 
                    # ifelse(mcav_DEG_GLM_treatment$logFC> 1 ,'Up','Down'),
                    # 'Stable')
```

```{r}
# set up data
ofav_DEG_GLM_treatment$expression_bonf = ifelse(ofav_DEG_GLM_treatment$PAdju_bonf < 0.05 & abs(ofav_DEG_GLM_treatment$logFC) >= 1, 
                     ifelse(ofav_DEG_GLM_treatment$logFC> 1 ,'Up','Down'),
                     'Stable')

#mcav_DEG_GLM_treatment$expression_bonf = ifelse(mcav_DEG_GLM_treatment$PAdju_bonf < 0.05 & abs(mcav_DEG_GLM_treatment$logFC) >= 1, 
                   #  ifelse(mcav_DEG_GLM_treatment$logFC> 1 ,'Up','Down'),
                   #  'Stable')
```
OFAV
```{r}
ofav_bonf_volc <- ggplot(data = ofav_DEG_GLM_treatment, 
            aes(x = logFC, 
                y = -log10(ofav_DEG_GLM_treatment$PValue), 
                colour=expression_bonf)) +
  geom_point(alpha=0.4, size=3.5) +
  scale_color_manual(values=c("blue", "grey","red"))+
  xlim(c(-4.5, 4.5)) +
  geom_vline(xintercept=c(-1,1),lty=4,col="black",lwd=0.8) +
  geom_hline(yintercept = 1.301,lty=4,col="black",lwd=0.8) +
  labs(x="log2(fold change)",
       y="-log10 (adj.p-value)",
       title="OFAV Differential expression - Bonferonni")  +
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5), 
        legend.position="right", 
        legend.title = element_blank())
```

```{r}
ofav_FDR_volc <- ggplot(data = ofav_DEG_GLM_treatment, 
            aes(x = logFC, 
                y = -log10(ofav_DEG_GLM_treatment$PValue), 
                colour=expression)) +
  geom_point(alpha=0.4, size=3.5) +
  scale_color_manual(values=c("blue", "grey","red"))+
  xlim(c(-4.5, 4.5)) +
  geom_vline(xintercept=c(-1,1),lty=4,col="black",lwd=0.8) +
  geom_hline(yintercept = 1.301,lty=4,col="black",lwd=0.8) +
  labs(x="log2(fold change)",
       y="-log10 (adj.p-value)",
       title="OFAV Differential expression - FDR")  +
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5), 
        legend.position="right", 
        legend.title = element_blank())
```

```{r}
#ggsave("ofav_FDR_volc.jpeg", plot=ofav_FDR_volc, width=8, height=8)
#ggsave("ofav_bonf_volc.jpeg", plot=ofav_bonf_volc, width=8, height=8)
```


MCAV
```{r}
ggplot(data = mcav_DEG_GLM_treatment, 
            aes(x = logFC, 
                y = -log10(mcav_DEG_GLM_treatment$PValue), 
                colour=expression_bonf)) +
  geom_point(alpha=0.4, size=3.5) +
  scale_color_manual(values=c("blue", "grey","red"))+
  xlim(c(-4.5, 4.5)) +
  geom_vline(xintercept=c(-1,1),lty=4,col="black",lwd=0.8) +
  geom_hline(yintercept = 1.301,lty=4,col="black",lwd=0.8) +
  labs(x="log2(fold change)",
       y="-log10 (adj.p-value)",
       title="MCAV Differential expression - Bonferonni")  +
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5), 
        legend.position="right", 
        legend.title = element_blank())
```

```{r}
ggplot(data = mcav_DEG_GLM_treatment, 
            aes(x = logFC, 
                y = -log10(mcav_DEG_GLM_treatment$PValue), 
                colour=expression)) +
  geom_point(alpha=0.4, size=3.5) +
  scale_color_manual(values=c("blue", "grey","red"))+
  xlim(c(-4.5, 4.5)) +
  geom_vline(xintercept=c(-1,1),lty=4,col="black",lwd=0.8) +
  geom_hline(yintercept = 1.301,lty=4,col="black",lwd=0.8) +
  labs(x="log2(fold change)",
       y="-log10 (adj.p-value)",
       title="MCAV Differential expression - FDR")  +
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5), 
        legend.position="right", 
        legend.title = element_blank())
```

# GO analysis
Ref: https://ucdavis-bioinformatics-training.github.io/2018-June-RNA-Seq-Workshop/friday/enrichment.html
```{r}
geneList <- ofav_annot_DEG_emap$PAdju_bonf
names(geneList) <- ofav_annot_DEG_emap$locus
ofav_mapping <- data.frame(ofav_annot_DEG_emap$locus, ofav_annot_DEG_emap$GOs)

# Create topGOData object -- not working
#GOdata <- new("topGOdata",
   # ontology = "CC",
  #  allGenes = geneList,
  #  geneSelectionFun = function(x)x,
  #  annot = annFUN.org, mapping = ofav_mapping)
```
GO using GO_MWU
https://github.com/z0on/GO_MWU/blob/master/GO_MWU.R

NOT WORKING 
```{r}
# Edit these to match your data file names: 
# two columns of comma-separated values: gene id, continuous measure of significance. To perform standard GO enrichment analysis based on Fisher's exact test, use binary measure (0 or 1, i.e., either sgnificant or not).
input= data.frame(ofav_annot_DEG_emap$locus, ofav_annot_DEG_emap$PAdju_bonf)

# two-column, tab-delimited, one line per gene, multiple GO terms separated by semicolon. If you have multiple lines per gene, use nrify_GOtable.pl prior to running this script.
goAnnotations <-

setwd("~/Desktop")
# download from http://www.geneontology.org/GO.downloads.ontology.shtml
goDatabase <- read.csv("go.obo")

# either MF, or BP, or CC
goDivision="CC"  

source("gomwu.functions.R")
```

```{r}
gomwuStats(input, goDatabase, goAnnotations, goDivision,
	perlPath="perl", # replace with full path to perl executable if it is not in your system's PATH already
	largest=0.1,  # a GO category will not be considered if it contains more than this fraction of the total number of genes
	smallest=5,   # a GO category should contain at least this many genes to be considered
	clusterCutHeight=0.25, # threshold for merging similar (gene-sharing) terms. See README for details.
#	Alternative="g" # by default the MWU test is two-tailed; specify "g" or "l" of you want to test for "greater" or "less" instead. 
#	Module=TRUE,Alternative="g" # un-remark this if you are analyzing a SIGNED WGCNA module (values: 0 for not in module genes, kME for in-module genes). In the call to gomwuPlot below, specify absValue=0.001 (count number of "good genes" that fall into the module)
#	Module=TRUE # un-remark this if you are analyzing an UNSIGNED WGCNA module 
)
```

