---
title: "ofav_mcav_sctld_count_import"
output: html_document
---

I always load all my packages as the first chunk in every rmarkdown I make. Its nicer than having them splayed around and also if you are knitting to a html then it means you will have no package dependency problems anywhere in the document. 

```{r Package Loading, include = F}
library(tximport)
library(DESeq2)
library(tidyverse)
library(tximport)
library(ComplexHeatmap)
#library(DEGreport)
library(limma)
library(edgeR)
library(pheatmap)
library(topGO)
library(ggplot2)
library(gridExtra)
library(reshape2)
#library(mixOmics)
library(RColorBrewer)
library(VennDiagram)
library(devtools)
library(factoextra)
library(topGO)
```


## Importing Salmon Quant Files

The quant files are all on the box 

**PLEASE DO NOT UPLOAD THESE TO ANY PUBLIC PLACES AS THIS IS THE RAW DATA**

We are going to download the quant files so you can make the count tables yourself. To do that we are going to have a folder on the desktop as this means you will not have to chnage any paths.You just need to make sure everything is saved in the correct folders 

```{bash}
cd ~/Desktop/sctld_jamboree
mkdir SCTLD_analysis
```

```{bash}
cd ~/Desktop/sctld_jamboree/SCTLD_analysis
mkdir scripts
```

Once you have done the above go to the box and download the  
+ ofav_salmon_quant/
+ mcav_salmon_quant/
+ metadata.csv

These should be saved in the SCTLD_analysis folder so you should have a directory tree as such 

Desktop/
      |
      ofav_salmon_quant/
      mcav_salmon_quant/
      scripts/
      metadata.csv
      

```{r Making Salmon Vectors Names ofav, include=FALSE}
#Loading quantification data output from the slippery Salmon
ofavall <- list.files(path = "~/Desktop/SCTLD_analysis/ofav_salmon_quant//", full.names = F, pattern = "\\_salmon$")

FILESofav <- file.path(ofavall, "quant.sf")

names(FILESofav) <- ofavall
head(FILESofav)
all(file.exists(FILESofav))
```

```{r Making Salmon Vectors Names mcav, include=FALSE}
#Loading quantification data output from the slippery Salmon
mcavall <- list.files(path = "~/Desktop/SCTLD_analysis/mcav_salmon_quant//", full.names = F, pattern = "\\_salmon$")

FILESmcav <- file.path(mcavall, "quant.sf")

names(FILESmcav) <- mcavall
head(FILESmcav)
all(file.exists(FILESmcav))
```

```{r Reading in Salmon Files ofav, include=F}
#Importing the count data using salmon quant.sf files and the text to gene file
setwd("~/Desktop/ofav_salmon_quant//")
txi.salmon.count_ofav <- tximport(FILESofav, type = "salmon", txOut = TRUE )
```

```{r Reading in Salmon Files mcav, include=F}
#Importing the count data using salmon quant.sf files and the text to gene file
setwd("~/Desktop/mcav_salmon_quant//")
txi.salmon.count_mcav <- tximport(FILESmcav, type = "salmon", txOut = TRUE )
```

We have read these in so lets just save all our count, abundance and length files as well

```{bash}
cd ~/Desktop
mkdir SCTLD_analysis
cd ~/Desktop/SCTLD_analysis/
mkdir count_folder
```

```{r writing csv for all info mcav, include = F}
#write.csv(txi.salmon.count_mcav$counts, file = "~/Desktop/SCTLD_analysis/count_folder/mcav_counts.csv")
#write.csv(txi.salmon.count_mcav$abundance , file = "~/Desktop/SCTLD_analysis/count_folder/mcav_abundance.csv")
#write.csv(txi.salmon.count_mcav$length, file = "~/Desktop/SCTLD_analysis/count_folder/mcav_length.csv")
```


```{r writing csv for all info ofav, include = F}
#write.csv(txi.salmon.count_ofav$counts, file = "~/Desktop/SCTLD_analysis/count_folder/ofav_counts.csv")
#write.csv(txi.salmon.count_ofav$abundance , file = "~/Desktop/SCTLD_analysis/count_folder/ofav_abundance.csv")
#write.csv(txi.salmon.count_ofav$length, file = "~/Desktop/SCTLD_analysis/count_folder/ofav_length.csv")
```

Now we are going to read in the metadata file and fix these for both species

```{r}
meta <- read.csv(file = "~/Desktop/metadata.csv")
str(meta)
```

Here, I filter the files I imported above and then filter them with the metadata file. This is needed as libraries failed and we only want the data for the samples that worked.  

```{r filter mcav}
meta %>% filter(salmon_label %in% mcavall) -> mcav_meta

## Check our salmon and meta lengths the same
length(mcavall)
nrow(mcav_meta)
```

```{r filter ofav}
meta %>% filter(salmon_label %in% ofavall) -> ofav_meta
length(ofavall)
nrow(ofav_meta)
```

## Read in Data and Make DESeq object
REF: http://www.nathalievialaneix.eu/doc/html/TP1_normalization.html 
Make a data frame of the counts for normalization with edgeR

```{r}
#setwd("~/Desktop")
#ofav_deunique <- read.csv("ofav_counts_deunique.csv")
#ofav_counts <- read.csv("ofav_counts_only_deunique.csv")
#ofav_lengths <- ofav_deunique$Length
```

```{r}
setwd("~/Desktop")
countdata <- read.delim("ofav_gene.counts.deuniqued", comment.char="#")
```
```{r}
ofav_lengths <- as.data.frame(countdata$Length)

# Set Gene ID's as row names
row.names(countdata) <- countdata$Geneid
# Remove first six columns (Geneid, chr, start, end, strand, length)
countdata <- countdata[ ,7:ncol(countdata)]
countdata <- countdata[ ,-4]

ofav_counts <- countdata
ofav_meta_dd <- ofav_meta[grep("K53|K54|K55|K57|K58|K59|K60|K61|K62|K63|K65|K66|K68|K69|K70|K71|K72|K73|K74|K75|K76|K78|K79|K80|K81|K82", ofav_meta$Working.Label), ]
```
```{r}
setwd("~/Desktop")
countdata_mc <- read.delim("mcav_counts", comment.char="#")
```

```{r}
mcav_lengths <- as.data.frame(countdata_mc$Length)

# Set Gene ID's as row names
row.names(countdata_mc) <- countdata_mc$Geneid
# Remove first six columns (Geneid, chr, start, end, strand, length)
countdata_mc <- countdata_mc[ ,7:ncol(countdata_mc)]

mcav_counts <- countdata_mc
```
```{r}
setwd("~/Desktop")
mcav_meta_dd <- read.csv("mcav_metadata_fixed.csv")
```


```{r Making DeSeq object, include=FALSE}
#making the dds model to use in deseq2
dds_mcav = DESeqDataSetFromMatrix(countdata_mc, mcav_meta_dd, ~ Correct_treatment)
dds_ofav = DESeqDataSetFromMatrix(countdata, ofav_meta_dd, ~ Correct_treatment)
```

```{r}
head(dds_mcav)
head(dds_ofav)
```
```{r}
# filter out low counts
keep_mcav <- rowSums(counts(dds_mcav)) >=10
keep_ofav <- rowSums(counts(dds_ofav)) >=10

dds_mcav_rmlow <- dds_mcav[keep_mcav,]
dds_ofav_rmlow <- dds_ofav[keep_ofav,]
```
How many non-zero genes?
```{r}
GeneCountsMcav <- counts(dds_mcav_rmlow)
idx.nz1 <- apply(GeneCountsMcav, 1, function(x) { all(x > 0)})
sum(idx.nz1)

GeneCountsOfav <- counts(dds_ofav_rmlow)
idx.nz2 <- apply(GeneCountsOfav, 1, function(x) { all(x > 0)})
sum(idx.nz2)
```

EdgeR used TMM normalization ()
```{r}
# filter out reads with count of 0
raw_counts_ofav <- ofav_counts[rowSums(ofav_counts) > 0,]
raw_counts_mcav <- mcav_counts[rowSums(mcav_counts) > 0, ]

# calucluate log transformed coutns
pseudo_counts_ofav <- log2(raw_counts_ofav + 1)
head(pseudo_counts_ofav)

pseudo_counts_mcav <- log2(raw_counts_mcav + 1)
head(pseudo_counts_mcav)

#ofav
df_raw_ofav <- melt(as.matrix(pseudo_counts_ofav), id = rownames(raw_counts_ofav))
names(df_raw_ofav)[1:2]<- c("id", "sample")
df_raw_ofav$method <- rep("Raw counts", nrow(df_raw_ofav))  

#mcav
df_raw_mcav <- melt(as.matrix(pseudo_counts_mcav), id = rownames(raw_counts_mcav))
names(df_raw_mcav)[1:2]<- c("id", "sample")
df_raw_mcav$method <- rep("Raw counts", nrow(df_raw_mcav))  
```
Visualize distribution of counts
```{r}
#ofav
df <- data.frame(rcounts = raw_counts_ofav[ ,1], prcounts = pseudo_counts_ofav[ ,1])

p <- ggplot(data=df, aes(x = rcounts, y = ..density..))
p <- p + geom_histogram(fill = "lightblue")
p <- p + theme_bw()
p <- p + ggtitle(paste0("count distribution ofav"))
p <- p + xlab("counts")

p2 <- ggplot(data=df, aes(x = prcounts, y = ..density..))
p2 <- p2 + geom_histogram(fill = "lightblue")
p2 <- p2 + theme_bw()
p2 <- p2 + ggtitle(paste0("count distribution ofav log2"))
p2 <- p2 + xlab(expression(log[2](counts + 1)))

#mcav
df2 <- data.frame(rcounts = raw_counts_mcav[ ,1], prcounts = pseudo_counts_mcav[ ,1])

p3 <- ggplot(data=df2, aes(x = rcounts, y = ..density..))
p3 <- p3 + geom_histogram(fill = "lightblue")
p3 <- p3 + theme_bw()
p3 <- p3 + ggtitle(paste0("count distribution mcav"))
p3 <- p3 + xlab("counts")

p4 <- ggplot(data=df2, aes(x = prcounts, y = ..density..))
p4 <- p4 + geom_histogram(fill = "lightblue")
p4 <- p4 + theme_bw()
p4 <- p4 + ggtitle(paste0("count distribution mcav log2"))
p4 <- p4 + xlab(expression(log[2](counts + 1)))

grid.arrange(p, p2, p3, p4, ncol = 2, nrow=2)
grid.arrange(p3, p4, ncol=2)
```
Look at variance and means because RNA data is overdispersed (i.e. variance in read counts will be higher than read counts)
```{r}
#ofav 
df <- data.frame(mean = apply(raw_counts_ofav[ ,ofav_meta_dd$Correct_treatment == "control"], 1, mean),
                 var = apply(raw_counts_ofav[ ,ofav_meta_dd$Correct_treatment == "control"], 1, var))
p <- ggplot(data=df, aes(x = mean, y = var))
p <- p + geom_point(colour = "orange")
p <- p + theme_bw()
p <- p + geom_abline(aes(intercept=0, slope=1))
p <- p + ggtitle("Variance versus mean in counts ofav") + ylab("variance")
print(p)

#mcav
df2 <- data.frame(mean = apply(raw_counts_mcav[ ,mcav_meta_dd$Correct_treatment == "control"], 1, mean),
                 var = apply(raw_counts_mcav[ ,mcav_meta_dd$Correct_treatment == "control"], 1, var))
p2 <- ggplot(data=df2, aes(x = mean, y = var))
p2 <- p2 + geom_point(colour = "orange")
p2 <- p2 + theme_bw()
p2 <- p2 + geom_abline(aes(intercept=0, slope=1))
p2 <- p2 + ggtitle("Variance versus mean in counts mcav") + ylab("variance")
print(p2)

#ofav 
df <- data.frame(mean = apply(raw_counts_ofav[ ,ofav_meta_dd$Correct_treatment == "control"], 1, mean),
                 var = apply(raw_counts_ofav[ ,ofav_meta_dd$Correct_treatment == "control"], 1, var))
df <- df[df$mean <= 5000, ]
p <- ggplot(data=df, aes(x = mean, y = var))
p <- p + geom_point(colour = "orange")
p <- p + theme_bw()
p <- p + geom_abline(aes(intercept=0, slope=1))
p <- p + ggtitle("Variance versus mean in counts ofav") + ylab("variance")
print(p)

#mcav
df2 <- data.frame(mean = apply(raw_counts_mcav[ ,mcav_meta_dd$Correct_treatment == "control"], 1, mean),
                 var = apply(raw_counts_mcav[ ,mcav_meta_dd$Correct_treatment == "control"], 1, var))
df2 <- df2[df2$mean <= 5000, ]
p2 <- ggplot(data=df2, aes(x = mean, y = var))
p2 <- p2 + geom_point(colour = "orange")
p2 <- p2 + theme_bw()
p2 <- p2 + geom_abline(aes(intercept=0, slope=1))
p2 <- p2 + ggtitle("Variance versus mean in counts mcav") + ylab("variance")
print(p2)
```
Do a basic PCA to determine sources of variance
```{r}
library(factoextra)
#ofav
res.pca <- prcomp(t(pseudo_counts_ofav), scale = TRUE)
fviz_eig(res.pca)
groups <- as.factor(ofav_meta_dd$Correct_treatment)
fviz_pca_ind(res.pca, col.ind=groups, label=FALSE)

#mcav
res.pca2 <- prcomp(t(pseudo_counts_mcav), scale = TRUE)
fviz_eig(res.pca2)
groups <- as.factor(mcav_meta_dd$Correct_treatment)
fviz_pca_ind(res.pca2, col.ind=groups, label=FALSE)
```
## Normalization ##
```{r}
# Use DESeq - corrects for library size NOT gene specific 
#ofav
sizefact <- estimateSizeFactorsForMatrix(raw_counts_ofav)
sizefact_dd <- estimateSizeFactors(dds_ofav)
deseq_normcount <- counts(sizefact_dd, normalized = TRUE)
test_normcount <- sweep(raw_counts_ofav, 2, sizefact, FUN = "/")
sum(test_normcount != deseq_normcount)

pseudo_deseq <- log2(deseq_normcount + 1)
df_deseq <- melt(pseudo_deseq, id = rownames(raw_counts_ofav))
names(df_deseq)[1:2]<- c("id", "sample")
df_deseq$method <- rep("DESeq (RLE)", nrow(df_deseq)) 

#mcav
sizefact2 <- estimateSizeFactorsForMatrix(raw_counts_mcav)
sizefact_dd2 <- estimateSizeFactors(dds_mcav)
deseq_normcount2 <- counts(sizefact_dd2, normalized = TRUE)
test_normcount2 <- sweep(raw_counts_mcav, 2, sizefact2, FUN = "/")
sum(test_normcount2 != deseq_normcount2)

pseudo_deseq2 <- log2(deseq_normcount2 + 1)
df_deseq2 <- melt(pseudo_deseq2, id = rownames(raw_counts_mcav))
names(df_deseq2)[1:2]<- c("id", "sample")
df_deseq2$method <- rep("DESeq (RLE)", nrow(df_deseq2)) 
```

```{r}
# EdgeR
#ofav
dge2 <- DGEList(raw_counts_ofav)
dge2

dge2$samples
dge2_keep <- filterByExpr(dge2, group=ofav_meta_dd$Correct_treatment, min.count=34)
summary(dge2_keep)
dge2 <- dge2[dge2_keep, , keep.lib.sizes=FALSE]

#mcav
dge3 <- DGEList(raw_counts_mcav)
dge3

dge3$samples
dge3_keep <- filterByExpr(dge3, group=mcav_meta_dd$Correct_treatment, min.count=15)
summary(dge3_keep)
dge3 <- dge3[dge3_keep, , keep.lib.sizes=FALSE]
```
Normalization method: total count
```{r}
#ofav
pseudo_TC <- log2(cpm(dge2) + 1)

df_TC <- melt(pseudo_TC, id = rownames(raw_counts_ofav))
names(df_TC)[1:2] <- c ("id", "sample")
df_TC$method <- rep("TC", nrow(df_TC))

# mcav
pseudo_TC2 <- log2(cpm(dge3) + 1)

df_TC2 <- melt(pseudo_TC2, id = rownames(raw_counts_mcav))
names(df_TC2)[1:2] <- c ("id", "sample")
df_TC2$method <- rep("TC", nrow(df_TC2))
```
Normalization method: reads per thousand counts (RPKM)
```{r}
#ofav
ofav_lengths_use <- ofav_lengths[rowSums(ofav_counts) > 0,]
pseudo_RPKM <- log2(rpkm(dge2, gene.length = ofav_lengths_use) + 1)

df_RPKM <- melt(as.matrix(pseudo_RPKM), id.vars = rownames(raw_counts_ofav))
names(df_RPKM)[1:2] <- c ("id", "sample")
df_RPKM$method <- rep("RPKM", nrow(df_RPKM))

#mcav
mcav_lengths_use <- mcav_lengths[rowSums(mcav_counts) > 0,]
pseudo_RPKM2 <- log2(rpkm(dge3, gene.length = mcav_lengths_use) + 1)

df_RPKM2 <- melt(as.matrix(pseudo_RPKM2), id = rownames(raw_counts_mcav))
names(df_RPKM2)[1:2] <- c ("id", "sample")
df_RPKM2$method <- rep("RPKM", nrow(df_RPKM2))
```

Normalization method: TMM
```{r}
#ofav
dge2 <- calcNormFactors(dge2, method = "TMM")
dge2$samples
dge2_keep <- filterByExpr(dge2, group=ofav_meta_dd$Correct_treatment, min.count=34)
dge2 <- dge2[dge2_keep, , keep.lib.sizes=FALSE]
pseudo_TMM <- log2(cpm(dge2) + 1)

df_TMM <- melt(as.matrix(pseudo_TMM), id = rownames(raw_counts_ofav))
names(df_TMM)[1:2] <- c ("id", "sample")
df_TMM$method <- rep("TMM", nrow(df_TMM))

#mcav
dge3 <- calcNormFactors(dge3, method = "TMM")
dge3$samples
dge3_keep <- filterByExpr(dge3, group=mcav_meta_dd$Correct_treatment, min.count=15)
dge3 <- dge3[dge3_keep, , keep.lib.sizes=FALSE]
pseudo_TMM2 <- log2(cpm(dge3) + 1)

df_TMM2 <- melt(as.matrix(pseudo_TMM2), id = rownames(raw_counts_mcav))
names(df_TMM2)[1:2] <- c ("id", "sample")
df_TMM2$method <- rep("TMM", nrow(df_TMM2))
```
Compare all normalization methods 
```{r, fig.width=10}
# Plots for ofav
df_allnorm <- rbind(df_raw_ofav, df_deseq, df_RPKM, df_TC, df_TMM)
df_allnorm$method <- factor(df_allnorm$method,
                            levels = c("Raw counts", "DESeq (RLE)", "RPKM","TC", "TMM"))

p <- ggplot(data=df_allnorm, aes(x=sample, y=value, fill=method))
p <- p + geom_boxplot()  
p <- p + theme_bw()
p <- p + ggtitle("Boxplots of normalized pseudo counts\n
for all samples by normalization methods - ofav")
p <- p + facet_grid(. ~ method) 
p <- p + ylab(expression(log[2] ~ (normalized ~ count + 1))) + xlab("")
p <- p + theme(title = element_text(size=10), axis.text.x = element_blank(), 
               axis.ticks.x = element_blank())
print(p)

ggsave("ofav_normalized_compare.jpeg", plot=p, width=10, height=8)
```

```{r, fig.width=10}
# Plots for mcav
df_allnorm2 <- rbind(df_raw_mcav, df_deseq2, df_RPKM2, df_TC2, df_TMM2)
df_allnorm2$method <- factor(df_allnorm2$method,
                            levels = c("Raw counts", "DESeq (RLE)", "RPKM", "TC", "TMM"))

p2 <- ggplot(data=df_allnorm2, aes(x=sample, y=value, fill=method))
p2 <- p2 + geom_boxplot()  
p2 <- p2 + theme_bw()
p2 <- p2 + ggtitle("Boxplots of normalized pseudo counts\n
for all samples by normalization methods - mcav")
p2 <- p2 + facet_grid(. ~ method) 
p2 <- p2 + ylab(expression(log[2] ~ (normalized ~ count + 1))) + xlab("")
p2 <- p2 + theme(title = element_text(size=10), axis.text.x = element_blank(), 
               axis.ticks.x = element_blank())
print(p2)

ggsave("mcav_normalized_compare.jpeg", plot=p2, width=10, height=8)
```

Density plots after normalization
```{r, fig.width=10}
#ofav
p <- ggplot(data=df_allnorm, aes(x=value, colour=sample))
p <- p + geom_density()  
p <- p + theme_bw()
p <- p + ggtitle("Density of normalized pseudo counts\n
for all samples by normalization methods - ofav\n
lines are colored by sample")
p <- p + facet_grid(. ~ method) 
p <- p + ylab(expression(log[2] ~ (normalized ~ count + 1))) + xlab("")
p <- p + theme(title = element_text(size=10), axis.text.x = element_blank(), 
               axis.ticks.x = element_blank())
p <- p + theme(legend.position = "none")
print(p)

ggsave("ofav_normalized_density.jpeg", plot=p, width=10, height=8)
```

```{r, fig.width=10}
#mcav
p2 <- ggplot(data=df_allnorm2, aes(x=value, colour=sample))
p2 <- p2 + geom_density()  
p2 <- p2 + theme_bw()
p2 <- p2 + ggtitle("Density of normalized pseudo counts\n
for all samples by normalization methods - mcav")
p2 <- p2 + facet_grid(. ~ method) 
p2 <- p2 + ylab(expression(log[2] ~ (normalized ~ count + 1))) + xlab("")
p2 <- p2 + theme(title = element_text(size=10), axis.text.x = element_blank(), 
               axis.ticks.x = element_blank())
p2 <- p2 + theme(legend.position = "none")
print(p2)

ggsave("mcav_normalized_density.jpeg", plot=p2, width=10, height=8)
```
# Repeat PCA with normalized TMM data

The red dashed line on the graph of PC variance explained bars indicates the expected average contribution. If the contribution of the variables were uniform, the expected value would be 1/length(variables) = 1/10 = 10%. For a given component, a variable with a contribution larger than this cutoff could be considered as important in contributing to the component.
```{r}
#ofav
res.pca <- prcomp(t(pseudo_TMM), scale = TRUE)
fviz_eig(res.pca)

# Contributions of variables to PC1
fviz_contrib(res.pca, choice = "var", axes = 1, top = 10)
# Contributions of variables to PC2
fviz_contrib(res.pca, choice = "var", axes = 2, top = 10)

fviz_pca_var(res.pca, select.var = list(contrib = 3))
groups <- as.factor(ofav_meta_dd$Correct_treatment)
fviz_pca_ind(res.pca, habillage = ofav_meta_dd$Correct_treatment, addEllipses=TRUE, label=FALSE)

groups <- as.factor(ofav_meta_dd$Project)
fviz_pca_ind(res.pca, habillage = ofav_meta_dd$percent_disease_progress, addEllipses=TRUE, label=FALSE)

groups <- as.factor(ofav_meta_dd$Exp_frag_genet)
fviz_pca_ind(res.pca, habillage = ofav_meta_dd$Exp_frag_genet, addEllipses=TRUE, label=FALSE)

groups <- as.factor(ofav_meta_dd$Exp_frag_genet)
fviz_pca_ind(res.pca, habillage = ofav_meta_dd$Exp_frag_genet, addEllipses=TRUE, label=FALSE)
```
The red dashed line on the graph of PC variance explained bars indicates the expected average contribution. If the contribution of the variables were uniform, the expected value would be 1/length(variables) = 1/10 = 10%. For a given component, a variable with a contribution larger than this cutoff could be considered as important in contributing to the component.
```{r}
#mcav
res.pca2 <- prcomp(t(pseudo_TMM2), scale = TRUE)
fviz_eig(res.pca2)

# Contributions of variables to PC1
fviz_contrib(res.pca2, choice = "var", axes = 1, top = 10)
# Contributions of variables to PC2
fviz_contrib(res.pca2, choice = "var", axes = 2, top = 10)

fviz_pca_var(res.pca2, select.var = list(contrib = 3))
groups <- as.factor(mcav_meta_dd$Correct_treatment)
fviz_pca_ind(res.pca2, habillage = mcav_meta_dd$Correct_treatment, addEllipses=TRUE, label=FALSE)

groups <- as.factor(mcav_meta_dd$Project)
fviz_pca_ind(res.pca2, habillage = mcav_meta_dd$Project, addEllipses=TRUE, label=FALSE)

groups <- as.factor(mcav_meta_dd$Exp_frag_genet)
fviz_pca_ind(res.pca2, habillage = mcav_meta_dd$Exp_frag_genet, addEllipses=TRUE, label=FALSE)
```
Try PCA with DESeq2 -- takes top ~500 rows
```{r}
vsd <- vst(dds_ofav, blind=FALSE)
plotPCA(vsd, intgroup=c("Correct_treatment", "Project"))

vsd2 <- vst(dds_mcav, blind=FALSE)
plotPCA(vsd2, intgroup=c("Correct_treatment","Project"))

# make it pretty
pcaData <- plotPCA(vsd, intgroup=c("Correct_treatment","Project"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=Correct_treatment, shape=Project)) +
   stat_ellipse() +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()+
   theme_bw()

pcaData2 <- plotPCA(vsd2, intgroup=c("Correct_treatment","Project","Exp_frag_genet"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData2, "percentVar"))
ggplot(pcaData2, aes(PC1, PC2, color=Correct_treatment, shape=Exp_frag_genet)) +
   stat_ellipse()+
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() +
   theme_bw()
```
## Differential Expression Analysis
BCV plots show biological variation. High variance (above red line) will reduce ability to detect DEG. Corrected p-values use FDR (Benjamini and Hochberg). A list of DEG's is generated using 4 different tests: pairwise exact test, GLM, GLM multivariate, and voom lm. GLM and voom are binomial estimate approaches and are more flexible for various experimental designs while the pairwise exact test is useful for one dimensional experiments. For this experiment, GLM will only tolerate up to two variables.

# OFAV
Pairwise exact test: DEG_pairwiseExact will contain list of DEG, pvals_pairwiseExact will contain list of raw and BH corrected p-vals.
```{r}
all_conditions <- setdiff(unique(ofav_meta_dd$Correct_treatment), "control")
DEG_pairwiseExact <- NULL
pvals_pairwiseExact <- NULL
par(mfrow = c(1,3))
for (ind in seq_along(all_conditions)) {
  # create dataset with 'wt' and current mutant and normalize (TMM)
  sel_cols <- union(grep("control", colnames(raw_counts_ofav)),
                    grep("control", colnames(raw_counts_ofav), invert=TRUE))
  cur_counts <- raw_counts_ofav[ ,sel_cols]
  group <- relevel(as.factor(ofav_meta_dd$Correct_treatment[sel_cols]), ref = "control")
  cur_dge <- DGEList(cur_counts, group = group)
  cur_dge <- calcNormFactors(cur_dge, method = "TMM")
  
  # estimate dispersions
  cur_dge <- estimateCommonDisp(cur_dge)
  cur_dge <- estimateTagwiseDisp(cur_dge)
  plotBCV(cur_dge, 
          main = paste0("BCV plot 'control' vs '", all_conditions[ind], "'"))
  
  # exact test
  res_et <- exactTest(cur_dge)
  cat("Exact test results:\n")
  print(res_et)
  pvals_pairwiseExact <- c(pvals_pairwiseExact, res_et$table$PValue,
                           p.adjust(res_et$table$PValue, method = "BH"))

  cat("Top 10 DEG for 'control' vs '", all_conditions[ind], "':\n")
  print(topTags(res_et))
  
  cur_res <- decideTestsDGE(res_et, adjust.method = "BH", p.value = 0.05)
  print(cur_res)
  sel_deg <- which(cur_res[ ,1] != 0)
  cur_res <- cbind(rownames(cur_counts)[sel_deg], cur_res[sel_deg,1],
                   rep(all_conditions[ind], length(sel_deg)))
  DEG_pairwiseExact <- rbind(DEG_pairwiseExact, cur_res)
}

DEG_pairwiseExact_ofav <- as.data.frame(DEG_pairwiseExact)
names(DEG_pairwiseExact_ofav) <- c("name", "UD", "condition")
listDEG_pairwiseExact_ofav <- unique(DEG_pairwiseExact_ofav$name)

pvals_pairwiseExact_ofav <- data.frame("pvalue" = pvals_pairwiseExact,
                                  "type" = rep(rep(c("raw", "adjusted"),
                                                   each = nrow(raw_counts_ofav)),
                                               length(all_conditions)),
                                  "condition" = rep(all_conditions,
                                                    each=nrow(raw_counts_ofav)*2))
```

```{r}
# view summary of pairwise exact results
#num unique DEGs = 15
table(DEG_pairwiseExact_ofav$condition)

# num up and down regulated (1 = upregulated)
table(DEG_pairwiseExact_ofav$condition, DEG_pairwiseExact_ofav$UD)

# total list length
length(listDEG_pairwiseExact_ofav)

# distribution of p-vals and adjusted p-vals
p <- ggplot(data = pvals_pairwiseExact_ofav, aes(x = pvalue, fill = type))
p <- p + geom_histogram()
p <- p + theme_bw()
p <- p + ggtitle("Histograms of raw/adjusted p-values for exact test")
p <- p + facet_grid(type ~ condition) 
p <- p + theme(title = element_text(size=10), axis.text.x = element_blank(), 
               axis.ticks.x = element_blank())
print(p)
```
GLM with group effect (recommended rather than the exact test because it is more robust to variation in experimental design while exact test is best for one dimensional experimental designs)
```{r}
cur_dge <- DGEList(raw_counts_ofav, group = ofav_meta_dd$Correct_treatment)
cur_dge <- calcNormFactors(cur_dge, method = "TMM")

group <- relevel(as.factor(ofav_meta_dd$Correct_treatment), ref = "control")
design_matrix <- model.matrix(~ group)
design_matrix <- design_matrix[,1:2]

cur_dge <- estimateGLMCommonDisp(cur_dge, design_matrix)
cur_dge <- estimateGLMTrendedDisp(cur_dge, design_matrix)
cur_dge <- estimateGLMTagwiseDisp(cur_dge, design_matrix)
plotBCV(cur_dge, main = paste0("BCV plot"))

# GLM fit
fit <- glmFit(cur_dge, design_matrix)
DEG_GLM1 <- NULL
pvals_GLM1 <- NULL
for (ind in 1) {
  res_GLM1 <- glmLRT(fit, coef = ind+1)
  pvals_GLM1 <- c(pvals_GLM1, res_GLM1$table$PValue,
                  p.adjust(res_GLM1$table$PValue, method = "BH"))

  cat("Top 10 DEG for 'control' vs '", all_conditions[ind], "':\n")
  print(topTags(res_GLM1))
  
  cur_res <- decideTestsDGE(res_GLM1, adjust.method = "BH", p.value = 0.05)
  print(cur_res)
  sel_deg <- which(cur_res[ ,1] != 0)
  cur_res <- cbind(rownames(raw_counts_ofav)[sel_deg], cur_res[sel_deg,1],
                   rep(all_conditions[ind], length(sel_deg)))
  DEG_GLM1 <- rbind(DEG_GLM1, cur_res)
}

DEG_GLM1_ofav <- as.data.frame(DEG_GLM1)
names(DEG_GLM1_ofav) <- c("name", "UD", "condition")
listDEG_GLM1_ofav <- unique(DEG_GLM1_ofav$name)

pvals_GLM1_ofav <- data.frame("pvalue" = pvals_GLM1,
                         "type" = rep(rep(c("raw", "adjusted"),
                                          each = nrow(raw_counts_ofav)),
                                      length(all_conditions)),
                         "condition" = rep(all_conditions,
                                           each=nrow(raw_counts_ofav)*2))

ofav_DEG_GLM_treatment <- as.data.frame(res_GLM1$table)
#FDR correction
ofav_DEG_GLM_treatment$PAdju_BH <- p.adjust(res_GLM1$table$PValue, method="BH")
#BONF correction
ofav_DEG_GLM_treatment$PAdju_bonf <- p.adjust(res_GLM1$table$PValue, method="bonferroni")

# save sign from FDR and BONF conrrections
# 1128 genes
ofav_DEG_GLM_treatment_sign_BH <- subset(ofav_DEG_GLM_treatment, ofav_DEG_GLM_treatment$PAdju_BH <0.05)
# 221 genes
ofav_DEG_GLM_treatment_sign_bonf <- subset(ofav_DEG_GLM_treatment, ofav_DEG_GLM_treatment$PAdju_bonf <0.05)
```

```{r}
# view summary of GLM results
#num unique DEGs = 1128  !!! Genes of interest for differentiating disease adn control !!!
table(DEG_GLM1_ofav$condition)

# num up and down regulated (1 = upregulated)
table(DEG_GLM1_ofav$condition, DEG_GLM1_ofav$UD)

# total list length
length(listDEG_GLM1_ofav)

# visualize distribution of p-vals and corrected p-vals
p <- ggplot(data = pvals_GLM1_ofav, aes(x = pvalue, fill = type))
p <- p + geom_histogram()
p <- p + theme_bw()
p <- p + ggtitle("Histogram of raw/adjusted p-values for GLM")
p <- p + facet_grid(type ~ condition) 
p <- p + theme(title = element_text(size=10), axis.text.x = element_blank(), 
               axis.ticks.x = element_blank())
print(p)
```
GLM with effects of treatment and project
```{r}
# create dataset with 'wt' and current mutant and normalize (TMM)
cur_dge <- DGEList(raw_counts_ofav, group = ofav_meta_dd$Correct_treatment)
cur_dge <- calcNormFactors(cur_dge, method = "TMM")

# create design matrix
design_matrix <- model.matrix(~ ofav_meta_dd$Correct_treatment + ofav_meta_dd$Exp_frag_genet)
design_matrix <- design_matrix[,1:2]

cur_dge <- estimateGLMCommonDisp(cur_dge, design_matrix)
cur_dge <- estimateGLMTrendedDisp(cur_dge, design_matrix)
cur_dge <- estimateGLMTagwiseDisp(cur_dge, design_matrix)
plotBCV(cur_dge, main = paste0("BCV plot"))

# GLM fit
fit <- glmFit(cur_dge, design_matrix)
DEG_GLM2 <- NULL
pvals_GLM2 <- NULL
for (ind in 1) {
  res_GLM2 <- glmLRT(fit, coef = ind+1)
  pvals_GLM2 <- c(pvals_GLM2, res_GLM2$table$PValue,
                  p.adjust(res_GLM2$table$PValue, method = "BH"))

  cat("Top 10 DEG for 'control' vs '", all_conditions[ind], "':\n")
  print(topTags(res_GLM2))
  
  cur_res <- decideTestsDGE(res_GLM2, adjust.method = "BH", p.value = 0.05)
  print(cur_res)
  sel_deg <- which(cur_res[ ,1] != 0)
  cur_res <- cbind(rownames(raw_counts_ofav)[sel_deg], cur_res[sel_deg,1],
                   rep(all_conditions[ind], length(sel_deg)))
  DEG_GLM2 <- rbind(DEG_GLM2, cur_res)
}

DEG_GLM2_ofav <- as.data.frame(DEG_GLM2)
names(DEG_GLM2_ofav) <- c("name", "UD", "condition")
listDEG_GLM2_ofav <- unique(DEG_GLM2_ofav$name)

pvals_GLM2_ofav <- data.frame("pvalue" = pvals_GLM2,
                         "type" = rep(rep(c("raw", "adjusted"),
                                          each = nrow(raw_counts_ofav)),
                                      length(all_conditions)),
                         "condition" = rep(all_conditions,
                                           each=nrow(raw_counts_ofav)*2))

ofav_DEG_GLM2_treatment <- as.data.frame(res_GLM2$table)
#FDR correction
ofav_DEG_GLM2_treatment$PAdju_BH <- p.adjust(res_GLM2$table$PValue, method="BH")
#BONF correction
ofav_DEG_GLM2_treatment$PAdju_bonf <- p.adjust(res_GLM2$table$PValue, method="bonferroni")

# save sign from FDR and BONF conrrections
# 1128 genes
ofav_DEG_GLM2_treatment_sign_BH <- subset(ofav_DEG_GLM2_treatment, ofav_DEG_GLM2_treatment$PAdju_BH <0.05)
# 221 genes
ofav_DEG_GLM2_treatment_sign_bonf <- subset(ofav_DEG_GLM2_treatment, ofav_DEG_GLM2_treatment$PAdju_bonf <0.05)
```

```{r}
# number of DEGs = 1128 with genotype and treatment
table(DEG_GLM2_ofav$condition)

table(DEG_GLM2_ofav$condition, DEG_GLM2_ofav$UD)

length(listDEG_GLM2_ofav)

p <- ggplot(data = pvals_GLM2_ofav, aes(x = pvalue, fill = type))
p <- p + geom_histogram()
p <- p + theme_bw()
p <- p + ggtitle("Histogram of raw/adjusted p-values for GLM multivariate")
p <- p + facet_grid(type ~ condition) 
p <- p + theme(title = element_text(size=10), axis.text.x = element_blank(), 
               axis.ticks.x = element_blank())
print(p)
```

Voom and LM test
```{r}
cur_dge <- DGEList(raw_counts_ofav)
cur_dge <- calcNormFactors(cur_dge, method = "TMM")
vdge <- voom(cur_dge, design_matrix, plot = TRUE)

# LM fit
fit <- lmFit(vdge, design_matrix)
fit <- eBayes(fit)
cur_gres <- decideTests(fit, adjust.method = "BH", p.value = 0.05)
head(cur_gres)

DEG_voom <- NULL
pvals_voom <- NULL
for (ind in 1) {
  res_voom <- fit$p.value
  pvals_voom <- c(pvals_voom, res_voom[ ,ind+1],
                  p.adjust(res_voom[ ,ind+1], method = "BH"))

  cat("Top 10 DEG for 'control' vs '", all_conditions[ind], "':\n")
  print(topTable(fit, coef = ind+1))
  
  sel_deg <- which(cur_gres[ ,ind+1] != 0)
  cur_res <- cbind(rownames(raw_counts_ofav)[sel_deg], 
                   cur_gres[sel_deg,ind+1],
                   rep(all_conditions[ind], length(sel_deg)))
  DEG_voom <- rbind(DEG_voom, cur_res)
}

DEG_voom_ofav <- as.data.frame(DEG_voom)
names(DEG_voom_ofav) <- c("name", "UD", "condition")
listDEG_voom_ofav <- unique(DEG_voom_ofav$name)

pvals_voom_ofav <- data.frame("pvalue" = pvals_voom,
                         "type" = rep(rep(c("raw", "adjusted"),
                                          each = nrow(raw_counts_ofav)),
                                      length(all_conditions)),
                         "condition" = rep(all_conditions,
                                           each=nrow(raw_counts_ofav)*2))
```

```{r}
# number of DEGs = 585
table(DEG_voom_ofav$condition)

table(DEG_voom_ofav$condition, DEG_voom_ofav$UD)

length(listDEG_voom_ofav)

p <- ggplot(data = pvals_voom_ofav, aes(x = pvalue, fill = type))
p <- p + geom_histogram()
p <- p + theme_bw()
p <- p + ggtitle("Histogram of raw/adjusted p-values for voom")
p <- p + facet_grid(type ~ condition) 
p <- p + theme(title = element_text(size=10), axis.text.x = element_blank(), 
               axis.ticks.x = element_blank())
print(p)
```

Overlapping Results -- Venn Diagram -- MCAV
```{r}
vd <- venn.diagram(x=list("Exact test" = listDEG_pairwiseExact_ofav,
                          "GLM\n treatment" = listDEG_GLM1_ofav, 
                          "GLM\n treatment + project" = listDEG_GLM2_ofav,
                          "voom" = listDEG_voom_ofav), 
                   fill = brewer.pal(4, "Set3"), 
                   cat.col = c("darkgreen", "black", "darkblue", "darkred"),
                   cat.cex = 1.5, fontface="bold", filename="venn_DEG_ofav.tiff")
```
# MCAV

Pairwise exact test: DEG_pairwiseExact will contain list of DEG, pvals_pairwiseExact will contain list of raw and BH corrected p-vals.
```{r}
all_conditions <- setdiff(unique(mcav_meta_dd$Correct_treatment), "control")
DEG_pairwiseExact <- NULL
pvals_pairwiseExact <- NULL
par(mfrow = c(1,3))
for (ind in seq_along(all_conditions)) {
  # create dataset with 'wt' and current mutant and normalize (TMM)
  sel_cols <- union(grep("control", colnames(raw_counts_mcav)),
                    grep("control", colnames(raw_counts_mcav), invert=TRUE))
  cur_counts <- raw_counts_mcav[ ,sel_cols]
  group <- relevel(as.factor(mcav_meta_dd$Correct_treatment[sel_cols]), ref = "control")
  cur_dge <- DGEList(cur_counts, group = group)
  cur_dge <- calcNormFactors(cur_dge, method = "TMM")
  
  # estimate dispersions
  cur_dge <- estimateCommonDisp(cur_dge)
  cur_dge <- estimateTagwiseDisp(cur_dge)
  plotBCV(cur_dge, 
          main = paste0("BCV plot 'control' vs '", all_conditions[ind], "'"))
  
  # exact test
  res_et <- exactTest(cur_dge)
  cat("Exact test results:\n")
  print(res_et)
  pvals_pairwiseExact <- c(pvals_pairwiseExact, res_et$table$PValue,
                           p.adjust(res_et$table$PValue, method = "BH"))

  cat("Top 10 DEG for 'control' vs '", all_conditions[ind], "':\n")
  print(topTags(res_et))
  
  cur_res <- decideTestsDGE(res_et, adjust.method = "BH", p.value = 0.05)
  print(cur_res)
  sel_deg <- which(cur_res[ ,1] != 0)
  cur_res <- cbind(rownames(cur_counts)[sel_deg], cur_res[sel_deg,1],
                   rep(all_conditions[ind], length(sel_deg)))
  DEG_pairwiseExact <- rbind(DEG_pairwiseExact, cur_res)
}

DEG_pairwiseExact_mcav <- as.data.frame(DEG_pairwiseExact)
names(DEG_pairwiseExact_mcav) <- c("name", "UD", "condition")
listDEG_pairwiseExact_mcav <- unique(DEG_pairwiseExact_mcav$name)

pvals_pairwiseExact_mcav <- data.frame("pvalue" = pvals_pairwiseExact,
                                  "type" = rep(rep(c("raw", "adjusted"),
                                                   each = nrow(raw_counts_mcav)),
                                               length(all_conditions)),
                                  "condition" = rep(all_conditions,
                                                    each=nrow(raw_counts_mcav)*2))
```

```{r}
# view summary of pairwise exact results
#num unique DEGs = 88
table(DEG_pairwiseExact_mcav$condition)

# num up and down regulated (1 = upregulated)
table(DEG_pairwiseExact_mcav$condition, DEG_pairwiseExact_mcav$UD)

# total list length
length(listDEG_pairwiseExact_mcav)

# distribution of p-vals and adjusted p-vals
p <- ggplot(data = pvals_pairwiseExact_mcav, aes(x = pvalue, fill = type))
p <- p + geom_histogram()
p <- p + theme_bw()
p <- p + ggtitle("Histograms of raw/adjusted p-values for exact test")
p <- p + facet_grid(type ~ condition) 
p <- p + theme(title = element_text(size=10), axis.text.x = element_blank(), 
               axis.ticks.x = element_blank())
print(p)
```
GLM with group effect (recommended rather than the exact test because it is more robust to variation in experimental design while exact test is best for one dimensional experimental designs)
```{r}
cur_dge <- DGEList(raw_counts_mcav, group = mcav_meta_dd$Correct_treatment)
cur_dge <- calcNormFactors(cur_dge, method = "TMM")

group <- relevel(as.factor(mcav_meta_dd$Correct_treatment), ref = "control")
design_matrix <- model.matrix(~ group)
design_matrix <- design_matrix[,1:2]

cur_dge <- estimateGLMCommonDisp(cur_dge, design_matrix)
cur_dge <- estimateGLMTrendedDisp(cur_dge, design_matrix)
cur_dge <- estimateGLMTagwiseDisp(cur_dge, design_matrix)
plotBCV(cur_dge, main = paste0("BCV plot"))

# GLM fit
fit <- glmFit(cur_dge, design_matrix)
DEG_GLM1 <- NULL
pvals_GLM1 <- NULL
for (ind in 1) {
  res_GLM1 <- glmLRT(fit, coef = ind+1)
  pvals_GLM1 <- c(pvals_GLM1, res_GLM1$table$PValue,
                  p.adjust(res_GLM1$table$PValue, method = "BH"))

  cat("Top 10 DEG for 'control' vs '", all_conditions[ind], "':\n")
  print(topTags(res_GLM1))
  
  cur_res <- decideTestsDGE(res_GLM1, adjust.method = "BH", p.value = 0.05)
  print(cur_res)
  sel_deg <- which(cur_res[ ,1] != 0)
  cur_res <- cbind(rownames(raw_counts_mcav)[sel_deg], cur_res[sel_deg,1],
                   rep(all_conditions[ind], length(sel_deg)))
  DEG_GLM1 <- rbind(DEG_GLM1, cur_res)
}

DEG_GLM1_mcav <- as.data.frame(DEG_GLM1)
names(DEG_GLM1_mcav) <- c("name", "UD", "condition")
listDEG_GLM1_mcav <- unique(DEG_GLM1_mcav$name)

pvals_GLM1_mcav <- data.frame("pvalue" = pvals_GLM1,
                         "type" = rep(rep(c("raw", "adjusted"),
                                          each = nrow(raw_counts_mcav)),
                                      length(all_conditions)),
                         "condition" = rep(all_conditions,
                                           each=nrow(raw_counts_mcav)*2))

mcav_DEG_GLM_treatment <- as.data.frame(res_GLM1$table)
#FDR correction
mcav_DEG_GLM_treatment$PAdju_BH <- p.adjust(res_GLM1$table$PValue, method="BH")
#BONF correction
mcav_DEG_GLM_treatment$PAdju_bonf <- p.adjust(res_GLM1$table$PValue, method="bonferroni")

# save sign from FDR and BONF conrrections
# 57 genes
mcav_DEG_GLM_treatment_sign_BH <- subset(mcav_DEG_GLM_treatment, mcav_DEG_GLM_treatment$PAdju_BH <0.05)
# 18 genes
mcav_DEG_GLM_treatment_sign_bonf <- subset(mcav_DEG_GLM_treatment, mcav_DEG_GLM_treatment$PAdju_bonf <0.05)
```

```{r}
# view summary of GLM results
#num unique DEGs = 57
table(DEG_GLM1_mcav$condition)

# num up and down regulated (1 = upregulated)
table(DEG_GLM1_mcav$condition, DEG_GLM1_mcav$UD)

# total list length
length(listDEG_GLM1_mcav)

# visualize distribution of p-vals and corrected p-vals
p <- ggplot(data = pvals_GLM1_mcav, aes(x = pvalue, fill = type))
p <- p + geom_histogram()
p <- p + theme_bw()
p <- p + ggtitle("Histogram of raw/adjusted p-values for GLM")
p <- p + facet_grid(type ~ condition) 
p <- p + theme(title = element_text(size=10), axis.text.x = element_blank(), 
               axis.ticks.x = element_blank())
print(p)
```
GLM with effects of treatment and project
```{r}
# create dataset with 'wt' and current mutant and normalize (TMM)
cur_dge <- DGEList(raw_counts_mcav, group = mcav_meta_dd$Correct_treatment)
cur_dge <- calcNormFactors(cur_dge, method = "TMM")

# create design matrix
design_matrix <- model.matrix(~ mcav_meta_dd$Project + mcav_meta_dd$Correct_treatment)
design_matrix <- design_matrix[,1:2]

cur_dge <- estimateGLMCommonDisp(cur_dge, design_matrix)
cur_dge <- estimateGLMTrendedDisp(cur_dge, design_matrix)
cur_dge <- estimateGLMTagwiseDisp(cur_dge, design_matrix)
plotBCV(cur_dge, main = paste0("BCV plot"))

# GLM fit
fit <- glmFit(cur_dge, design_matrix)
DEG_GLM2 <- NULL
pvals_GLM2 <- NULL
for (ind in 1) {
  res_GLM2 <- glmLRT(fit, coef = ind+1)
  pvals_GLM2 <- c(pvals_GLM2, res_GLM2$table$PValue,
                  p.adjust(res_GLM2$table$PValue, method = "BH"))

  cat("Top 10 DEG for 'control' vs '", all_conditions[ind], "':\n")
  print(topTags(res_GLM2))
  
  cur_res <- decideTestsDGE(res_GLM2, adjust.method = "BH", p.value = 0.05)
  print(cur_res)
  sel_deg <- which(cur_res[ ,1] != 0)
  cur_res <- cbind(rownames(raw_counts_mcav)[sel_deg], cur_res[sel_deg,1],
                   rep(all_conditions[ind], length(sel_deg)))
  DEG_GLM2 <- rbind(DEG_GLM2, cur_res)
}

DEG_GLM2_mcav <- as.data.frame(DEG_GLM2)
names(DEG_GLM2_mcav) <- c("name", "UD", "condition")
listDEG_GLM2_mcav <- unique(DEG_GLM2_mcav$name)

pvals_GLM2_mcav <- data.frame("pvalue" = pvals_GLM2,
                         "type" = rep(rep(c("raw", "adjusted"),
                                          each = nrow(raw_counts_mcav)),
                                      length(all_conditions)),
                         "condition" = rep(all_conditions,
                                           each=nrow(raw_counts_mcav)*2))
```

```{r}
# number of DEGs = 7
table(DEG_GLM2_mcav$condition)

table(DEG_GLM2_mcav$condition, DEG_GLM2_mcav$UD)

length(listDEG_GLM2_mcav)

p <- ggplot(data = pvals_GLM2_mcav, aes(x = pvalue, fill = type))
p <- p + geom_histogram()
p <- p + theme_bw()
p <- p + ggtitle("Histogram of raw/adjusted p-values for GLM multivariate")
p <- p + facet_grid(type ~ condition) 
p <- p + theme(title = element_text(size=10), axis.text.x = element_blank(), 
               axis.ticks.x = element_blank())
print(p)
```

Voom and LM test
```{r}
cur_dge <- DGEList(raw_counts_mcav)
cur_dge <- calcNormFactors(cur_dge, method = "TMM")
vdge <- voom(cur_dge, design_matrix, plot = TRUE)

# LM fit
fit <- lmFit(vdge, design_matrix)
fit <- eBayes(fit)
cur_gres <- decideTests(fit, adjust.method = "BH", p.value = 0.05)
head(cur_gres)

DEG_voom <- NULL
pvals_voom <- NULL
for (ind in 1) {
  res_voom <- fit$p.value
  pvals_voom <- c(pvals_voom, res_voom[ ,ind+1],
                  p.adjust(res_voom[ ,ind+1], method = "BH"))

  cat("Top 10 DEG for 'control' vs '", all_conditions[ind], "':\n")
  print(topTable(fit, coef = ind+1))
  
  sel_deg <- which(cur_gres[ ,ind+1] != 0)
  cur_res <- cbind(rownames(raw_counts_mcav)[sel_deg], 
                   cur_gres[sel_deg,ind+1],
                   rep(all_conditions[ind], length(sel_deg)))
  DEG_voom <- rbind(DEG_voom, cur_res)
}

DEG_voom_mcav <- as.data.frame(DEG_voom)
names(DEG_voom_mcav) <- c("name", "UD", "condition")
listDEG_voom_mcav <- unique(DEG_voom_mcav$name)

pvals_voom_mcav <- data.frame("pvalue" = pvals_voom,
                         "type" = rep(rep(c("raw", "adjusted"),
                                          each = nrow(raw_counts_mcav)),
                                      length(all_conditions)),
                         "condition" = rep(all_conditions,
                                           each=nrow(raw_counts_mcav)*2))
```

```{r}
# number of DEGs = 6
table(DEG_voom_mcav$condition)

table(DEG_voom_mcav$condition, DEG_voom_mcav$UD)

length(listDEG_voom_mcav)

p <- ggplot(data = pvals_voom_mcav, aes(x = pvalue, fill = type))
p <- p + geom_histogram()
p <- p + theme_bw()
p <- p + ggtitle("Histogram of raw/adjusted p-values for voom")
p <- p + facet_grid(type ~ condition) 
p <- p + theme(title = element_text(size=10), axis.text.x = element_blank(), 
               axis.ticks.x = element_blank())
print(p)
```

Overlapping Results -- Venn Diagram -- MCAV
```{r}
vd <- venn.diagram(x=list("Exact test" = listDEG_pairwiseExact_mcav,
                          "GLM\n treatment" = listDEG_GLM1_mcav, 
                          "GLM\n treatment + project" = listDEG_GLM2_mcav,
                          "voom" = listDEG_voom_mcav), 
                   fill = brewer.pal(4, "Set3"), 
                   cat.col = c("darkgreen", "black", "darkblue", "darkred"),
                   cat.cex = 1.5, fontface="bold", filename="venn_DEG_mcav.tiff")
```

# Removing the effect of Project (i.e. location, mote vs. smithsonian)
```{r}
# create dataset with 'wt' and current mutant and normalize (TMM)
cur_dge <- DGEList(raw_counts_ofav, group = ofav_meta_dd$Correct_treatment)
cur_dge <- calcNormFactors(cur_dge, method = "TMM")

# create design matrix
design_matrix <- model.matrix(~ ofav_meta_dd$Correct_treatment - ofav_meta_dd$Project)
design_matrix <- design_matrix[,1:2]

cur_dge <- estimateGLMCommonDisp(cur_dge, design_matrix)
cur_dge <- estimateGLMTrendedDisp(cur_dge, design_matrix)
cur_dge <- estimateGLMTagwiseDisp(cur_dge, design_matrix)
plotBCV(cur_dge, main = paste0("BCV plot"))

# GLM fit
fit <- glmFit(cur_dge, design_matrix)
DEG_GLM2 <- NULL
pvals_GLM2 <- NULL
for (ind in 1) {
  res_GLM2 <- glmLRT(fit, coef = ind+1)
  pvals_GLM2 <- c(pvals_GLM2, res_GLM2$table$PValue,
                  p.adjust(res_GLM2$table$PValue, method = "BH"))

  cat("Top 10 DEG for 'control' vs '", all_conditions[ind], "':\n")
  print(topTags(res_GLM2))
  
  cur_res <- decideTestsDGE(res_GLM2, adjust.method = "BH", p.value = 0.05)
  print(cur_res)
  sel_deg <- which(cur_res[ ,1] != 0)
  cur_res <- cbind(rownames(raw_counts_ofav)[sel_deg], cur_res[sel_deg,1],
                   rep(all_conditions[ind], length(sel_deg)))
  DEG_GLM2 <- rbind(DEG_GLM2, cur_res)
}

DEG_GLM2_noproj_ofav <- as.data.frame(DEG_GLM2)
names(DEG_GLM2_noproj_ofav) <- c("name", "UD", "condition")
listDEG_GLM2_noproj_ofav <- unique(DEG_GLM2_noproj_ofav$name)

pvals_GLM2_noproj_ofav <- data.frame("pvalue" = pvals_GLM2,
                         "type" = rep(rep(c("raw", "adjusted"),
                                          each = nrow(raw_counts_ofav)),
                                      length(all_conditions)),
                         "condition" = rep(all_conditions,
                                           each=nrow(raw_counts_ofav)*2))
```

```{r}
# create dataset with 'control' and 'current mutant'diseased' and normalize (TMM)
cur_dge <- DGEList(raw_counts_mcav, group = mcav_meta_dd$Correct_treatment)
cur_dge <- calcNormFactors(cur_dge, method = "TMM")

# create design matrix
design_matrix <- model.matrix(~ mcav_meta_dd$Correct_treatment - mcav_meta_dd$Project)
design_matrix <- design_matrix[,1:2]

cur_dge <- estimateGLMCommonDisp(cur_dge, design_matrix)
cur_dge <- estimateGLMTrendedDisp(cur_dge, design_matrix)
cur_dge <- estimateGLMTagwiseDisp(cur_dge, design_matrix)
plotBCV(cur_dge, main = paste0("BCV plot"))

# GLM fit
fit <- glmFit(cur_dge, design_matrix)
DEG_GLM2 <- NULL
pvals_GLM2 <- NULL
for (ind in 1) {
  res_GLM2 <- glmLRT(fit, coef = ind+1)
  pvals_GLM2 <- c(pvals_GLM2, res_GLM2$table$PValue,
                  p.adjust(res_GLM2$table$PValue, method = "BH"))

  cat("Top 10 DEG for 'control' vs '", all_conditions[ind], "':\n")
  print(topTags(res_GLM2))
  
  cur_res <- decideTestsDGE(res_GLM2, adjust.method = "BH", p.value = 0.05)
  print(cur_res)
  sel_deg <- which(cur_res[ ,1] != 0)
  cur_res <- cbind(rownames(raw_counts_mcav)[sel_deg], cur_res[sel_deg,1],
                   rep(all_conditions[ind], length(sel_deg)))
  DEG_GLM2 <- rbind(DEG_GLM2, cur_res)
}

DEG_GLM2_noproj_mcav <- as.data.frame(DEG_GLM2)
names(DEG_GLM2_noproj_mcav) <- c("name", "UD", "condition")
listDEG_GLM2_noproj_mcav <- unique(DEG_GLM2_noproj_mcav$name)

pvals_GLM2_noproj_mcav <- data.frame("pvalue" = pvals_GLM2,
                         "type" = rep(rep(c("raw", "adjusted"),
                                          each = nrow(raw_counts_mcav)),
                                      length(all_conditions)),
                         "condition" = rep(all_conditions,
                                           each=nrow(raw_counts_mcav)*2))
```

Overlapping Results -- Venn Diagram -- OFAV
```{r}
# make the correct DFs from code above first
vd <- venn.diagram(x=list("Exact test" = listDEG_pairwiseExact_ofav,
                          "GLM\n treatment" = listDEG_GLM1_ofav, 
                          "GLM\n treatment + project" = listDEG_GLM2_ofav,
                          "GLM\n treatment - project" = listDEG_GLM2_noproj_ofav,
                          "voom" = listDEG_voom_ofav), 
                   fill = brewer.pal(5, "Set3"), 
                   cat.col = c("darkgreen", "black", "darkblue", "darkred", "orange"),
                   cat.cex = 1.5, fontface="bold", filename="venn_DEG_ofav2.tiff")
```

Overlapping Results -- Venn Diagram -- MCAV
```{r}
# make the correct DFs from code above first
vd <- venn.diagram(x=list("Exact test" = listDEG_pairwiseExact_mcav,
                          "GLM\n treatment" = listDEG_GLM1_mcav, 
                          "GLM\n treatment + project" = listDEG_GLM2_mcav,
                          "GLM\n treatment - project" = listDEG_GLM2_noproj_mcav,
                          "voom" = listDEG_voom_mcav), 
                   fill = brewer.pal(5, "Set3"), 
                   cat.col = c("darkgreen", "black", "darkblue", "darkred", "orange"),
                   cat.cex = 1.5, fontface="bold", filename="venn_DEG_mcav2.tiff")
```

```{r}
# save p-value tables of DEG from GLM (our favorite model) with treatment

#write.csv(ofav_DEG_GLM_treatment, "ofav_DEG_GLM_treatment.csv")
#write.csv(mcav_DEG_GLM_treatment, "mcav_DEG_GLM_treatment.csv")
```
